_posts/2017-11-10-choosing-a-cni-provider.md

```diff
@@ -0,0 +1,315 @@
+      <td>Yes</td>
+      <td>No</td>
+      <td>Yes*</td>
+      <td>No</td>
+      <td>Yes</td>
+      <td>No</td>
+    </tr>
+    <tr>
+      <td>flannel</td>
+      <td>vxlan</td>
+      <td>No</td>
+      <td>No</td>
+      <td>No</td>
+      <td>No</td>
+      <td>No</td>
+      <td>No</td>
+      <td>No</td>
+    </tr>
+    <tr>
+      <td>kopeio-vxlan</td>
+      <td>vxlan*</td>
+      <td>No</td>
+      <td>No</td>
+      <td>No</td>
+      <td>No</td>
+      <td>No</td>
+      <td>No</td>
+      <td>No</td>
+    </tr>
+    <tr>
+      <td>kube-router</td>
+      <td>layer3</td>
+      <td>Yes</td>
+      <td>Yes</td>
+      <td>No</td>
+      <td>No</td>
+      <td>No</td>
+      <td>No</td>
+      <td>No</td>
+    </tr>
+    <tr>
+      <td>romana</td>
+      <td>layer3</td>
+      <td>Yes</td>
+      <td>Yes</td>
+      <td>No</td>
+      <td>Yes</td>
+      <td>No</td>
+      <td>Yes</td>
+      <td>Yes</td>
+    </tr>
+    <tr>
+      <td>weave</td>
+      <td>vxlan*</td>
+      <td>No</td>
+      <td>Yes</td>
+      <td>Yes</td>
+      <td>No</td>
+      <td>Yes</td>
+      <td>Yes</td>
+      <td>Yes</td>
+    </tr>
+  </tbody>
+</table>
+
+### Table Details
+
+#### Network Model
+
+The Network Model with providers is either encapsulated networking such as VXLAN
+or unencapsulated layer 2 networking.  Encapsulating network traffic requires
+compute to process, so theoretically is slower.  In my opinion, most use cases
+will not be impacted by the overhead.  More about VXLAN on
+[wikipedia](https://en.wikipedia.org/wiki/Virtual_Extensible_LAN).
+
+#### BGP
+
+Border Gateway Protocol, BGP, is another nice to have a feature with CNI, if you
+need it.  It is an exterior gateway protocol designed to exchange routing and
+reachability information on the Internet.  BGP can assist with pod to pod
+networking between clusters.
+
+#### Network Policies
+
+A kubernetes.io blog post about network policies in 1.8
+[here](http://blog.kubernetes.io/2017/10/enforcing-network-policies-in-kubernetes.html).
+
+> Kubernetes now offers functionality to enforce rules about which pods can
+communicate with each other using network policies. This feature is has become
+stable Kubernetes 1.7 and is ready to use with supported networking plugins. The
+Kubernetes 1.8 release has added better capabilities to this feature.
+
+#### Mesh Networking
+
+This feature allows for "pod to pod" networking between Kubernetes clusters.
+This technology is not Kubernetes federation, but it pure networking between
+pods.
+
+#### Encyption
+
+Encrypting the network control plane, so all TCP and UDP traffic is encrypted.
+
+#### Ingress / Egress Policies
+
+The network policies are both Kubernetes and Non-Kubernetes routing control. For
+instance, many providers will allow an administrator to block a pod
+communicating with an EC2 instance meta and data service on 169.254.169.254.
+
+## Summary
+
+Pick one.  Make a decision quickly, don't spend days deciding, and test with
```

Comment:
Don't rule out kubenet.  For most users, it remains the best choice.

Reply:
Agreed